scipy/spatial/transform/_rotation.pyx

```diff
@@ -223,6 +236,124 @@ cdef double[:, :] _compute_euler_from_matrix(
 
     return angles
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef double[:, :] _compute_euler_from_quat(
+    np.ndarray[double, ndim=2] quat, const uchar[:] seq, bint extrinsic=False
+):
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by
+    # 1. Replacing "d" with "-d" for intrinsic rotation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+    
+    if not extrinsic:
+        seq = seq[::-1]
+
+    cdef int i = _elementary_basis_index(seq[0])
+    cdef int j = _elementary_basis_index(seq[1])
+    cdef int k = _elementary_basis_index(seq[2])
+
+    cdef bint is_proper = i == k
+    if is_proper:
+        k = 3 - i - j # get third axis
+        
+    # Step 0
+    # Check if permutation is even (+1) or odd (-1)     
+    cdef int sign = (i - j) * (j - k) * (k - i) // 2
+
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+
+    # some forward definitions
+    cdef double[:, :] angles = _empty2(num_rotations, 3)
+    cdef double[:] _angles # accessor for each rotation
+    cdef double a, b, c, d
+    cdef double half_sum, half_diff
+    cdef double eps = 1e-7
+    cdef bint safe1, safe2, safe, adjust
+
+    for ind in range(num_rotations):
+        _angles = angles[ind, :]
+
+        # Step 1
+        # Permutate quaternion elements            
+        if is_proper:
+            a = quat[ind, 3]
+            b = quat[ind, i]
+            c = quat[ind, j]
+            d = quat[ind, k] * sign
+        else:
+            a = quat[ind, 3] - quat[ind, j]
+            b = quat[ind, i] + quat[ind, k] * sign
+            c = quat[ind, j] + quat[ind, 3]
+            d = quat[ind, k] * sign - quat[ind, i]
+        
+        # Step 2
+        # Compute second angle...
+        _angles[1] = 2*atan2(sqrt(c*c + d*d), sqrt(a*a + b*b))
```

Comment:
With all the advantages mentioned (apparent better precision, no need to compute norm, divide and guard against out of bounds values) this formula looks like the way to go.

I came to the conclusion that in numerical code `atan2` should always be preferred when it is possible to get sin and cos independently.

The algorithm looks super clean now, I believe it will become a "golden standard" to convert quaternion to Euler angles.

Reply:
Awesome! I'm glad you agree :)

Reply:
There is a C-math function `hypot` to compute `sqrt(x*x + y*y)`. It is a preferred way to compute it. 

Perhaps makes no difference here (as the numbers are close, etc.) but still a good practice.

It should be accessible as `from libc.math cimport hypot`.

Reply:
Oh thanks for the tip, I did not know that! :)